src/getChildEventSubscriber.js

```diff
@@ -0,0 +1,137 @@
+      case 'didFocus':
+        return didFocusSubscribers;
+      case 'willBlur':
+        return willBlurSubscribers;
+      case 'didBlur':
+        return didBlurSubscribers;
+      default:
+        return null;
+    }
+  };
+
+  const emit = (eventName, payload) => {
+    const subscribers = getChildSubscribers(eventName);
+    subscribers &&
+      subscribers.forEach(subs => {
+        // $FlowFixMe - Payload should probably understand generic state type
+        subs(payload);
+      });
+  };
+
+  let isSelfFocused = false;
+
+  const cleanup = () => {
+    upstreamSubscribers.forEach(subs => subs && subs.remove());
+  };
+
+  const upstreamEvents = [
+    'willFocus',
+    'didFocus',
+    'willBlur',
+    'didBlur',
+    'action',
+  ];
+
+  const upstreamSubscribers = upstreamEvents.map(eventName => {
+    addListener(eventName, (payload: NavigationEventPayload) => {
+      const { state, lastState, action } = payload;
+      const lastFocusKey = lastState && lastState.routes[lastState.index].key;
+      const focusKey = state && state.routes[state.index].key;
+
+      const isFocused = focusKey === key;
+      const wasFocused = lastFocusKey === key;
+      const lastRoute =
+        lastState && lastState.routes.find(route => route.key === key);
+      const newRoute = state && state.routes.find(route => route.key === key);
+      const childPayload = { state: newRoute, lastState: lastRoute, action };
+
+      const didNavigate =
+        (lastState && lastState.isTransitioning) !==
+        (state && state.isTransitioning);
+
+      const isTransitioning = !!state && state.isTransitioning;
+      const wasTransitioning = !!lastState && lastState.isTransitioning;
+      const didStartTransitioning = !wasTransitioning && isTransitioning;
+      const didFinishTransitioning = wasTransitioning && !isTransitioning;
+
+      if (eventName !== 'action') {
+        switch (eventName) {
+          case 'didFocus':
+            isSelfFocused = true;
+            break;
+          case 'didBlur':
+            isSelfFocused = false;
+            break;
+        }
+        emit(eventName, childPayload);
+        return;
+      }
+
+      // now we're exclusively handling the "action" event
+
+      if (newRoute) {
+        // fire this event to pass navigation events to children subscribers
+        emit('action', childPayload);
+      }
+      if (isFocused && didStartTransitioning && !isSelfFocused) {
+        emit('willFocus', childPayload);
+      }
+      if (isFocused && didFinishTransitioning && !isSelfFocused) {
+        emit('didFocus', childPayload);
+        isSelfFocused = true;
+      }
+      if (!isFocused && didStartTransitioning && isSelfFocused) {
+        emit('willBlur', childPayload);
+      }
+      if (!isFocused && didFinishTransitioning && isSelfFocused) {
+        emit('didBlur', childPayload);
+        isSelfFocused = false;
+      }
+    });
+  });
+
+  return (eventName, eventHandler) => {
```

Comment:
this is where we would handle lazily adding upstream subscribers, and we'd augment the `remove` function to cleanup

Reply:
yeah, exactly.

Fixed in https://github.com/react-navigation/react-navigation/issues/3354